src/formatter/simple.rs

```diff
@@ -1,361 +1,433 @@
+    pub fn new(buffer: T) -> Simple<T> {
         Simple {
-            buf: buf,
+            buffer: buffer,
             name_stack: vec![],
-            failures: vec![],
+            failed: vec![],
         }
     }
 
-    fn failures_summary(&self) -> String {
-        let res = String::with_capacity(100);
-        let mut idx = 0;
-        self.failures
-            .iter()
-            .map(|fail| {
-                idx += 1;
-                format!("  {}) {}\n", idx, fail)
-            })
-            .fold(res, |mut acc, elt| {
-                acc.push_str(&elt);
-                acc
-            })
+    fn padding(depth: usize) -> String {
+        "  ".repeat(depth)
     }
 
-    fn write_summary(&mut self, result: runner::RunnerResult) -> String {
-        let (res, report) = match result {
-            Ok(report) => ("ok", report),
-            Err(report) => ("FAILED", report),
-        };
-
-        format!("\n\ntest result: {}. {} examples; {} passed; {} failed;",
-                res,
-                report.total_tests,
-                report.success_count,
-                report.error_count)
+    fn enter_suite(&mut self, info: &SuiteInfo) {
+        self.name_stack.push(ScopeInfo::Suite(info.clone()));
+        let _ = writeln!(self.buffer, "\nrunning tests");
+        let label: &str = info.label.into();
+        let _ = writeln!(self.buffer, "{} {:?}:", label, info.name);
     }
-}
 
-impl<'a, T: io::Write> EventHandler for Simple<'a, T> {
-    fn trigger(&mut self, event: &Event) {
-        // FIXME: do something with the io::Error ?
-        let _ = match *event {
-            Event::StartRunner => writeln!(self.buf, "Running tests:\n"),
-            Event::StartDescribe(ref name) |
-            Event::StartTest(ref name) => {
-                self.name_stack.push(name.clone());
-                Ok(())
-            }
-            Event::EndTest(result) => {
-                if result.is_err() && !self.name_stack.is_empty() {
-                    let failure_name = self.name_stack.join(" | ");
-                    self.failures.push(failure_name);
+    fn exit_suite(&mut self, report: &SuiteReport) {
+        let _ = writeln!(self.buffer, "\nfailures:");
+        let failed = mem::replace(&mut self.failed, vec![]);
+        for scope_stack in failed {
+            for (indent, scope) in scope_stack.into_iter().enumerate() {
+                match scope {
+                    ScopeInfo::Suite(info) => {
+                        let padding = Self::padding(indent);
+                        let label: &str = info.label.into();
+                        let _ = writeln!(self.buffer, "{}{} {:?}:", padding, label, info.name);
+                    }
+                    ScopeInfo::Context(info) => {
+                        let padding = Self::padding(indent);
+                        let label: &str = info.label.into();
+                        let _ = writeln!(self.buffer, "{}{} {:?}:", padding, label, info.name);
```

Comment:
I believe you said multiple times the same thing, as there is duplication in this code. I say it because I see twice the same code. :) This needs to be factorized

Reply:
There is no formatter `formatter:Simple` anymore. It is called `formatter::serial::Formatter` now and has a corresponding `formatter::parallel::Formatter` equivalent.

The former prints output to the console as events come in, while the latter waits for all tests in a suite to have finished, before printing anything.

We might want to pass the config to the formatter through the `EventHandler` API to allow the implementation of a smart wrapper type that intelligently chooses the proper strategy for a given `Configuration`. We might end up having to do so anyway, as we might want to be able to configure whether to show timing information in the log, or not, etc. 🤔